Include/internal/pycore_code.h

```diff
@@ -463,6 +463,34 @@ adaptive_counter_backoff(uint16_t counter) {
 }
 
 
+/* Line array cache for tracing */
+
+extern int _PyCode_CreateLineArray(PyCodeObject *co);
+
+static inline int
+_PyCode_InitLineArray(PyCodeObject *co)
+{
+    if (co->_co_linearray) {
+        return 0;
+    }
+    return _PyCode_CreateLineArray(co);
+}
+
+static inline int
+_PyCode_LineNumberFromArray(PyCodeObject *co, int index)
+{
+    assert(co->_co_linearray != NULL);
+    assert(index >= 0);
+    if (co->_co_linearray_entry_size == 2) {
+        return ((int16_t *)co->_co_linearray)[index];
+    }
+    else {
+        assert(co->_co_linearray_entry_size == 4);
+        return ((int32_t *)co->_co_linearray)[index];
+    }
+}
```

Comment:
Why are we using signed values here? Line numbers can't be negative, can they?

Reply:
Yes. Missing line numbers are represented as -1.

Reply:
@markshannon Can a magic constant `#define NO_LINE_NUMBER (uint32_t )-1` be used here instead? I see no complexity difference between `if (line < 0)` and `if (line == NO_LINE_NUMBER)`.

Reply:
It would be a reasonable design, but it would be a large change spanning the compiler and interpreter.